### PR TITLE
Added support for complex input for Cholesky decomposition

### LINKED ISSUE #44637: Complex-valued Cholesky decomposition
## 🚀 Feature
Support for complex-valued `torch.cholesky`.

## Motivation
Pytorch 1.6 supports complex-valued tensors, however, most of the matrix operations cannot be written in terms of their real analogue  (in contrast to e.g. multiplication, which currently also does not support complex-valued data). Cholesky decomposition is important in decorrelating e.g. Monte Carlo data. The MAGMA backend also supports Cholesky decompositions for complex-valued data.

## Pitch
torch.cholesky supports complex-valued data.

## Alternatives
Possible workarounds

cc @ezyang @gchanan @zou3519 @anjali411 @dylanbespalko @nikitaved

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Cholesky decomposition now works for complex inputs.

Fixes #44637.

### CODE REVIEW

File: torch/testing/_internal/common_utils.py [hunk 1]
```diff
@@ -1535,11 +1535,20 @@ def random_fullrank_matrix_distinct_singular_value(matrix_size, *batch_dims,
...
 
     A = torch.randn(batch_dims + (matrix_size, matrix_size), dtype=dtype, device=device)
     u, _, v = A.svd()
-    s = torch.arange(1., matrix_size + 1, dtype=dtype, device=device).mul_(1.0 / (matrix_size + 1)).diag()
-    return u.matmul(s.expand(batch_dims + (matrix_size, matrix_size)).matmul(v.transpose(-2, -1)))
+    if dtype.is_complex:
+        real_dtype = torch.float32 if dtype is torch.complex64 else torch.float64
+    else:
+        real_dtype = dtype
+    s = ((1. - matrix_size) * torch.rand(matrix_size, dtype=real_dtype, device=device) + matrix_size).diag()
```
Comment (anjali411) [inline torch/testing/_internal/common_utils.py:1546]: can you explain mathematically what's going on here and why it's the correct thing to do?

File: torch/testing/_internal/common_utils.py [hunk 2]
```diff
@@ -1535,11 +1535,20 @@ def random_fullrank_matrix_distinct_singular_value(matrix_size, *batch_dims,
     silent = kwargs.get("silent", False)
     if silent and not torch._C.has_lapack:
         return torch.ones(matrix_size, matrix_size, dtype=dtype, device=device)
+    symmetric = kwargs.get("symmetric", False)
 
     A = torch.randn(batch_dims + (matrix_size, matrix_size), dtype=dtype, device=device)
     u, _, v = A.svd()
-    s = torch.arange(1., matrix_size + 1, dtype=dtype, device=device).mul_(1.0 / (matrix_size + 1)).diag()
-    return u.matmul(s.expand(batch_dims + (matrix_size, matrix_size)).matmul(v.transpose(-2, -1)))
+    if dtype.is_complex:
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply torch/testing/_internal/common_utils.py:1542]: Yes, right now this function is used only for reals. I introduced this change before trying to generate real and imag components separately. When svd and matmul both support complex, it won't be needed here to have `s` as reals, so I'll delete this.

File: torch/testing/_internal/common_utils.py (hunk 1 above)
Comment (IvanYashchuk) [reply torch/testing/_internal/common_utils.py:1546]: The random singular values (samples from a uniform distribution with values between 1.0 and `matrix_size`) being used for constructing the random matrix. Previously it was deterministic, so it just introduces more randomness to the test suit. Looking at it now, I see that this change is maybe unnecessary and not related to this PR. I am going to revert this change, since tests pass well on many runs with or without randomness.

**Comment by IvanYashchuk** [comment]
@anjali411, regarding your comments about changes `random_fullrank_matrix_distinct_singular_value`. Now I think those changes are unnecessary and unrelated to current PR, so I removed them, and use `random_symmetric_pd_matrix` tests pass well.

File: test/test_torch.py [hunk 3]
```diff
@@ -7745,40 +7745,57 @@ def cholesky_test_helper(n, batchsize, device, upper):
...
 
         def cholesky_test_helper(n, batch_dims, upper):
-            A = random_symmetric_pd_matrix(n, *batch_dims, dtype=dtype, device=device)
+            # This is a workaround while there is no support for complex random_symmetric_pd_matrix
+            if dtype.is_complex:
+                real_dtype = torch.float32 if dtype is torch.complex64 else torch.float64
+                A_real = random_symmetric_pd_matrix(n, *batch_dims, dtype=real_dtype, device=device)
+                A_imag = random_symmetric_pd_matrix(n, *batch_dims, dtype=real_dtype, device=device)
+                A = A_real + 1j * A_imag
+                A = 0.5 * (A + A.transpose(-2, -1).conj())
```
Comment (ngimel) [inline test/test_torch.py:7759]: This makes imaginary part 0 (A is symmetric matrix, A.transpose(-2,-1) is equal to A, A+A.conj() kills imaginary part), so the tests aren't testing much. You want hermitian positive definite input.

Comment (IvanYashchuk) [reply test/test_torch.py]: I've fixed this.

File: test/test_torch.py [hunk 4]
```diff
@@ -7743,41 +7743,68 @@ def cholesky_test_helper(n, batchsize, device, upper):
...
+            (random_symmetric_pd_matrix,
+             random_fullrank_matrix_distinct_singular_value)
+
+        # This is a workaround while there is no support for complex random_symmetric_pd_matrix
+        if dtype.is_complex:
+            real_dtype = torch.float32 if dtype is torch.complex64 else torch.float64
+            A_real = random_fullrank_matrix_distinct_singular_value(10, dtype=real_dtype, device=device)
+            A_imag = random_fullrank_matrix_distinct_singular_value(10, dtype=real_dtype, device=device)
+            A = A_real + 1j * A_imag
+            A = A @ A.t().conj()
```
Comment (anjali411) [inline test/test_torch.py:7791]: @ ?

**Review by anjali411 — APPROVED** [review]

**Review by anjali411 — COMMENTED** [review]
We should also update the backward definition for cholesky operation and gradcheck test for complex. @IvanYashchuk would you like to create a follow-up PR for that? you can find the gradient formula for cholesky decomposition in https://arxiv.org/pdf/1701.00392.pdf 4.53 and 4.54

**Comment by IvanYashchuk** [comment]
> We should also update the backward definition for cholesky operation and gradcheck test for complex. @IvanYashchuk would you like to create a follow-up PR for that? you can find the gradient formula for cholesky decomposition in https://arxiv.org/pdf/1701.00392.pdf 4.53 and 4.54

Yes, I'll do that.
